subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingVariantSelector.java

```diff
@@ -101,130 +96,78 @@ public ResolvedArtifactSet select(ResolvedVariantSet producer, Factory factory)
+        differentTransforms.add(last);
 
-        for (int i = 0; i < candidates.size(); i++) {
-            if (hasBetterMatch[i]) {
-                continue;
-            }
-            boolean candidateIsDifferent = true;
-            Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant> current = candidates.get(i);
-            for (int j = i + 1; j < candidates.size(); j++) {
-                if (hasBetterMatch[j]) {
-                    continue;
-                }
-                int index = j; // Needed to use inside lambda below
-                candidateIsDifferent = compareCandidates(matcher, current, candidates.get(index)).map(candidate -> {
-                    if (candidate != current) {
-                        // The other is better, current is not part of result
-                        return false;
-                    } else {
-                        // The other is disambiguated by current, never consider other again
-                        hasBetterMatch[index] = true;
-                    }
-                    return true;
-                }).orElse(true);
-            }
-            if (candidateIsDifferent) {
-                shortestTransforms.add(current);
+        // Find any other candidate which does not match with the last candidate.
+        for (int i = 0; i < candidates.size() - 1; i++) {
+            TransformedVariant current = candidates.get(i);
+            if (candidatesDifferent(matcher, current, last)) {
+                differentTransforms.add(current);
             }
         }
-        return shortestTransforms;
-    }
 
-    private List<Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant>> disambiguateWithSchema(AttributeMatcher matcher, List<Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant>> candidates, ImmutableAttributes componentRequested, AttributeMatchingExplanationBuilder explanationBuilder) {
-        List<AttributeContainerInternal> candidateAttributes = candidates.stream().map(pair -> pair.getRight().attributes).collect(Collectors.toList());
-        List<AttributeContainerInternal> matches = matcher.matches(candidateAttributes, componentRequested, explanationBuilder);
-        if (matches.size() == 1) {
-            AttributeContainerInternal singleMatch = matches.get(0);
-            return candidates.stream().filter(pair -> pair.getRight().attributes.equals(singleMatch)).collect(Collectors.toList());
-        } else if (matches.size() > 0 && matches.size() < candidates.size()) {
-            // We know all are compatibles, so this is only possible if some disambiguation happens but not getting us to 1 candidate
-            return candidates.stream().filter(pair -> matches.contains(pair.getRight().attributes)).collect(Collectors.toList());
-        }
-        return candidates;
+        return differentTransforms.size() == candidates.size() ? candidates : differentTransforms;
```

Comment:
When doesn't this return the same thing as the contents of `differentTransforms`?

`differentTransforms.size() > candidates.size()` cannot happen AFAIK.
`differentTransforms.size() < candidates.size()` returns differentTransforms.
`differentTransforms.size() == candidates.size()` returns candidates, but this should be the same as differentTransforms?

Reply:
They are the same contents. Though if the lists are identical I figured it would be better to return the original list instead of a copy of the original list. If you think this is superfluous this can be simplified to `return differentTransforms`.

Reply:
Yeah, I think just returning `differentTransforms` would be better, since the extra conditional here makes it seem like there's something special going on.